Assets/Scripts/Communication/CommunicationManager.cs

```diff
@@ -0,0 +1,392 @@
+	 * Private communication helper logic
+	 ********************************************/
+	private static void Connect()
+	{
+		Communication.Connect(new Uri(url));
+	}
+
+	private static void Disconnect()
+	{
+		Communication.Close();
+	}
+
+	private static bool IsUnauthenticated(Dictionary<string, object> response)
+	{
+		if (!response.ContainsKey("Success")){
+			return false;
+		}
+
+		bool success = (bool)response["Success"];
+		if (!success)
+		{
+			string error = response["Error"].ToString();
+			if (error.Contains("User is not authenticated"))
+			{
+				Debug.Log("User is not authenticated, retry logging in.");
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	// Used to login to server with cached session token
+	public static bool RetryLogin()
+	{
+		// Create the request, decrypt session token, and send it
+		var request = new Dictionary<string, object>();
+		string _encryptedToken = PlayerPrefs.GetString("session");
+		string _loginToken = AES.Decrypt(_encryptedToken, GenerateAESKey());
+		request["Command"] = "LGN";
+		request["token"] = _loginToken;
+		Communication.SendString(Json.ToString(request));
+		// Wait for the response, then parse
+		string strResponse = null;
+		while (strResponse == null)
+		{
+			strResponse = Communication.RecvString();
+		}
+		var response = Json.ToDict(strResponse);
+		Debug.Log("Response to Retry Login: " + response.ToString());
+		// Error Handling
+		bool success = (bool)response["Success"];
+		if (success)
+		{
+			Debug.Log("user re-logged in with token: " + _loginToken);
+		}
+		else
+		{
+			Debug.Log("error logging user in with existing token");
+			PlayerPrefs.DeleteKey("session");
+			PlayerPrefs.Save();
+		}
+		return success;
+	}
+
+	private static bool SendCommand(Dictionary<string, object> request)
+	{
+		// Verify the websocket is still connected and try to reconnect if it isn't
+		if (!Communication.IsConnected())
+		{
+			Communication.Close();
+			Communication.Connect(new Uri(url));
+		}
+
+		// Failed to reconnect with the server
+		if (!Communication.IsConnected())
+		{
+			return false;
+		}
+
+		// Send the request
+		Communication.SendString(Json.ToString(request));
+
+		return true;
+	}
+
+	private static Dictionary<string, object> GetNextResponse(){
+		// Wait for the response
+		string strResponse = null;
+		Dictionary<string, object> response = null;
+
+		int retryCount = 0;
+		const int maxRetries = 20;
+		while (strResponse == null && retryCount < maxRetries)
+		{
+			strResponse = Communication.RecvString();
+			if (strResponse != null)
+			{
+				response = Json.ToDict(strResponse);
+				if (IsUnauthenticated(response))
+				{
+					// Server says we are not logged in, re-authenticate
+					RetryLogin();
+					strResponse = null;
+					response = null;
+					throw new Exception("Response indicates user was unauthenticated please try again");
+				}
+
+
+			}
+
+			if (strResponse == null)
+			{
+				retryCount++;
+				Thread.Sleep(100);
```

Comment:
I think this should either be lowered (10 ms?) or also could benefit from the notify system.